README.md

```diff
@@ -108,354 +108,354 @@ in order to add your own service into Lucida. Let's break it down into two steps
+      and if it can handle text,
+      the types your service should handle are `text`, `url`, and `unlearn`, 
+      and if it can handle image,
+      the types your service should handle are `image` and `unlearn`.
+      See step 2 for details on how to specify the type of knowledge that your service can learn.
+      If `type` is `text` or `url`, `data[i]` is the `i`th piece of text or url as new knowledge
+      and `tags[i]` is the id of the `i`th piece of text generated by a hash function in the command center;
+      if `type` is `image`, `data[i]` is the `i`th image as new knowledge
+      (notice that it is the actual string representation of an image and thus can very long),
+      and `tags[i]` is the label/name of the `i`th image received from the front end;
+      if `type` is `unlearn`, `data` should be ignored by your service (usually a list of an empty string),
+      and `tags[i]` is the `i`th id of the text or the image label to delete,
+      based on whether the service can handle text or image.
+
+      * If the function call is `infer`:
+
+      Each `QueryInput` in `content` corresponds to one service (CMD is not considered to be a service)
+      in the service graph, a connected directed acyclic graph (DAG) describing all services that are needed for the query.
+      Thus, for the following service graph, two `QueryInput`s are present in `content`, each being a node in the graph:
+
+      ```Command Center (CMD) -> Your Own Service 0 (YOS0) -> Your Own Service 1 (YOS1)```
+
+      For `QueryInput`, `type` can be `text` for plain text, or `image` for image (no `url` for `infer`).
+      See step 2 for details on how to specify the type of query that your service can process.
+      If `type` is `text`, `data[0]` is the `0`th piece of text;
+      if `type` is `image`, `data[0]` is the `0`th image. There is only one string in `data`.
+      `tags` have different meanings from `learn` and are of the following format:
+
+      ```
+      [host, port, <size of the following list>, <list of integers>]
+      ```
+
+      . `tags` in the `i`th `QueryInput` in `content` describe the location of the `i`th node and its relation to other nodes.
+      By location, we mean that the `host:port` specifies the location of the `i`th node.
+      By relation to other nodes, we mean that the list of integers specifies the indices of nodes that the `i`th node points to.
+
+      Therefore, the above service graph results in a `QuerySpec` that looks like this:
+
+      ```
+      { name: "query", 
+      content: [ 
+      { type: "text",
+      data: [ "What's the speed of light?" ],
+      tags: ["localhost", "8083", "1", "1"] },
+      { type: "text",
+      data: [ "What's the speed of light?" ],
+      tags: ["localhost", "9090", "0"] } ] }
+      ```
+
+      . We can define arbitrarily complicated service graphs. For example, for the following service graph:
```

Comment:
These periods don't seem like they belong. Is there another way to format these lines that start with periods and commas?